### PR TITLE
Bounding `ElectionProvider`

### PR DESCRIPTION
**Pull request by georgesdib**
Implements **Step 1** of #9511

### CODE REVIEW

File: frame/election-provider-multi-phase/src/lib.rs [hunk 1]
```diff
@@ -1037,9 +1077,11 @@ pub mod pallet {
 			let maybe_max_voters = maybe_max_voters.map(|x| x as usize);
 			let maybe_max_targets = maybe_max_targets.map(|x| x as usize);
 
+			// TODO: is it right to pass 0 remaining?
 			let supports = T::GovernanceFallback::elect_with_bounds(
 				maybe_max_voters.unwrap_or(Bounded::max_value()),
 				maybe_max_targets.unwrap_or(Bounded::max_value()),
+				0,
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/lib.rs:1084]: should be `DataProvider::msp()`.

If we do a single page election, we want to make sure that we are using the "best" subset of voters, which is always `msp`.

File: frame/election-provider-multi-phase/src/lib.rs [hunk 2]
```diff
@@ -435,13 +443,29 @@ impl<C: Default> Default for RawSolution<C> {
...
+	DefaultNoBound,
+	TypeInfo,
+	MaxEncodedLen,
+)]
+#[codec(mel_bound(AccountId: MaxEncodedLen, MaxWinnersPerPage: Get<u32>, MaxBackersPerWinner: Get<u32>))]
+#[scale_info(skip_type_params(MaxWinnersPerPage, MaxBackersPerWinner))]
+pub struct ReadySolution<
+	AccountId: Eq + Clone + sp_std::fmt::Debug,
+	MaxWinnersPerPage: Get<u32>,
+	MaxBackersPerWinner: Get<u32>,
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/lib.rs:462]: You might as well make `ReadySolution` be generic over `T: Config`.

File: frame/election-provider-multi-phase/src/lib.rs [hunk 3]
```diff
@@ -450,6 +474,12 @@ pub struct ReadySolution<A> {
 	pub compute: ElectionCompute,
 }
 
+pub type ReadySolutionOf<T> = ReadySolution<
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/lib.rs:477]: Won't be needed if we make `ReadySolution<T: Config>`

File: frame/election-provider-multi-phase/src/lib.rs [hunk 4]
```diff
@@ -315,15 +317,21 @@ impl<T: Config> ElectionProvider for NoFallback<T> {
 	type BlockNumber = T::BlockNumber;
 	type DataProvider = T::DataProvider;
 	type Error = &'static str;
+	type MaxBackersPerWinner = T::MaxBackersPerWinner;
+	type MaxWinnersPerPage = T::MaxWinnersPerPage;
 
-	fn elect() -> Result<Supports<T::AccountId>, Self::Error> {
+	fn elect(_: PageIndex) -> Result<BoundedSupportsOf<Self>, Self::Error> {
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/lib.rs:323]: the `_: PageIndex` is not needed for this PR. That's step 2 in #9511.

Comment (kianenigma) [reply frame/election-provider-multi-phase/src/lib.rs:323]: In this PR, we only want to bound stuff, not paginate anything.

**Review by kianenigma — COMMENTED** [review]
Looks in the right direction, but please: 

1. Don't paginate anything in this PR, only bounding the return type of `elect`.
2. Make this PR be based on top of #11105 to have a cleaner diff. You can select this at the top of the page in the github UI.

**Comment by georgesdib** [comment]
Closing this PR as I have rewritten it in #11242

**Pull request closed without merging**